text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,163 @@
+
+For example:
+
+- `sntyu_a1b2c3...` - identifies a **user** API token
+- `sntya_a1b2c3...` - identifies an **API Application** token
+
+### Backend Token Hygiene
+
+API tokens should be stored in the database with a cryptographically secure hash (minimum SHA-256). Salting is not necessary since these values are sufficient in length.
+
+An additional column identifying the API token version will be added in order to correctly identify legacy API tokens.
+
+### Secret Verification Service
+
+In order to integrate with Github's Secret Scanning service, we will need to have an available service for them to submit potential matches to.
+
+More details about Github's secret scanning program can be found [here](https://docs.github.com/en/developers/overview/secret-scanning-partner-program).
+
+At a high-level, we will need to be able to:
+
+- Receive a list of one or more potential tokens
+- [Verify the signature of the request](https://docs.github.com/en/developers/overview/secret-scanning-partner-program#implement-signature-verification-in-your-secret-alert-service)
+- Determine if the token is a true positive or a false positive
+  - send this feedback back to Github
+- Disable/revoke the compromised token
+- Send a notification to the appropriate user
+
+> :memo: This secret verification service is not an immediate need or a dependency of implementing the improved API tokens.
+
+# Background
+
+Accidental credential leaks [happen](https://github.com/ChALkeR/notes/blob/master/Do-not-underestimate-credentials-leaks.md). Even though we provide several tools for users to limit the storage of sensitive information in Sentry it can still happen.
+
+Our support for various authentication methods (including 2FA, SSO, and SAML) helps mitigate access of potentially sensitive information, but API tokens cannot support these additional authentication gates.
+
+We can help customers further protect their accounts and data by providing a means of auto-detecting leaked API tokens.
+
+# Options Considered
+
+## Option #1
+
+The migration from the legacy tokens to the new tokens should be rolled out over X releases. This allows for a smooth transition of self-hosted instances keeping pace with the versions.
+
+> With `n` being the current version.
+
+### `n+1`
+
+- Add a Django migration including the `hashed_token`, `name`, and `token_version` fields.
+  - columns should allow `null` values for now.
+- Remove displaying the plaintext token in the frontend, and only display newly created tokens once.
+- Implement new logic to generate the new token pattern.
+  - The `token_version` should be set as `2`.
+- Implement a backwards compatible means of verifying legacy tokens. This should:
+  - Verify the legacy token is valid/invalid.
+  - If valid, hash the legacy token and store it in the table.
+  - Remove the plaintext representation of the token in the table.
+  - Mark the `token_version` as `1`.
```

Comment:
This seems like a task we could do incrementally as tokens are used. During authentication we know if a token is v1, and could either replace the token synchronously or dispatch a task to update the token.

Reply:
Yes, exactly! 😄 I'll try to re-word this to make it more clear.

Reply:
I re-wrote this whole section to clarify. Let me know your thoughts. :)